macho/input-files.cc

```diff
@@ -418,11 +418,18 @@ LoadCommand *ObjectFile<E>::find_load_command(Context<E> &ctx, u32 type) {
 template <typename E>
 Subsection<E> *
 ObjectFile<E>::find_subsection(Context<E> &ctx, u32 secidx, u32 addr) {
-  Subsection<E> *ret = nullptr;
-  for (Subsection<E> *subsec : subsections)
-    if (subsec->isec.secidx == secidx && subsec->input_addr <= addr)
-      ret = subsec;
-  return ret;
+  assert(subsections.size() > 0);
+  auto begin = subsections.begin();
+  auto end = subsections.end();
+  auto upper = std::upper_bound(begin, end, addr,
+      [](u32 addr, Subsection<E> *subsec) { return addr < subsec->input_addr; });
+
+  if (upper == begin)
+    return nullptr;
```

Comment:
```suggestion
  auto upper = std::ranges::upper_bound(subsections, addr, {}, &Subsection<E>::input_addr);

  if (upper == subsections.begin())
    return nullptr;
```

C++20 Ranges & Projections.

Reply:
I want to try this, but clang (with whatever headers are shipped with MacOSX12.3.sdk) replies: 

```
error: no member named 'upper_bound' in namespace 'std::ranges'; did you mean 'std::upper_bound'?
  auto upper = std::ranges::upper_bound(subsections, addr, {}, &Subsection<E>::input_addr);
               ^~~~~~~~~~~~~~~~~~~~~~~~
               std::upper_bound
```

Reply:
Sigh, looks like it’s a moment where Apple clang is lagging far behind (per https://en.cppreference.com/w/cpp/compiler_support). Nevermind then.